File: RxTango/python/examples/stats.py
"""Collect N samples, compute min/mean/max/std (one shot).

Polls double_scalar, accumulates N values, then prints statistics and exits.

Mirrors `TangoTestStats.java`.

Usage:
    python stats.py [device] [n-samples] [interval-ms]

    device       defaults to tango://localhost:10000/sys/tg_test/1
    n-samples    number of samples; defaults to 10
    interval-ms  poll interval; defaults to 500
"""

import asyncio
import math
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


async def main() -> None:
    device      = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    n_samples   = int(sys.argv[2]) if len(sys.argv) > 2 else 10
    interval_ms = int(sys.argv[3]) if len(sys.argv) > 3 else 500

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    done      = asyncio.Event()

    print(f"Collecting {n_samples} samples from {device}  (interval={interval_ms} ms) …\n")

    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.flat_map(lambda _: read_attribute(device, "double_scalar")),
        ops.take(n_samples),
        ops.to_list(),
    ).subscribe(
        on_next=lambda samples: (
            print(f"  n      = {len(samples)}"),
            print(f"  min    = {min(samples):+.6f}"),
            print(f"  max    = {max(samples):+.6f}"),
            print(f"  mean   = {sum(samples)/len(samples):+.6f}"),
            print(f"  std    = {math.sqrt(sum((x - sum(samples)/len(samples))**2 for x in samples) / len(samples)):+.6f}"),
        ),
        on_error=lambda e: (print(f"  ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=done.set,
        scheduler=scheduler,
    )

    await asyncio.wait_for(done.wait(), timeout=60.0)


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/stats.cpp
/**
 * Collect N samples, compute batch statistics, then exit.
 *
 * Mirrors Python's stats.py and Java's TangoTestStats.java.
 *
 * Usage:
 *   ./stats [device] [attribute] [N] [interval-ms]
 *   defaults: sys/tg_test/1  double_scalar  20  200
 */

#include <algorithm>
#include <chrono>
#include <condition_variable>
#include <cmath>
#include <iostream>
#include <mutex>
#include <numeric>
#include <string>
#include <vector>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

int main(int argc, char* argv[]) {
    const std::string device      = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr        = argc > 2 ? argv[2] : "double_scalar";
    const int         N           = argc > 3 ? std::stoi(argv[3]) : 20;
    const int         interval_ms = argc > 4 ? std::stoi(argv[4]) : 200;

    std::cout << "Collecting " << N << " samples of " << device << "/" << attr << " ...\n\n";

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([device, attr](long) {
            return rxtango::read_attribute<double>(device, attr);
        })
        .take(N)
        .to_vector()
        .subscribe(
            [N](std::vector<double> samples) {
                double sum  = std::accumulate(samples.begin(), samples.end(), 0.0);
                double mean = sum / N;
                double var  = 0.0;
                for (double x : samples) var += (x - mean) * (x - mean);
                var /= (N - 1);

                std::cout << "  N      = " << N << "\n"
                          << "  min    = " << *std::min_element(samples.begin(), samples.end()) << "\n"
                          << "  max    = " << *std::max_element(samples.begin(), samples.end()) << "\n"
                          << "  mean   = " << mean << "\n"
                          << "  stddev = " << std::sqrt(var) << "\n";
            },
            [&](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            },
            [&]() { std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one(); }
        );

    std::unique_lock<std::mutex> lk(m);
    cv.wait(lk, [&]{ return done; });
    return 0;
}
